src/System.Management.Automation/engine/parser/PSType.cs

```diff
@@ -436,15 +437,29 @@ private Type GetBaseTypes(Parser parser, TypeDefinitionAst typeDefinitionAst, ou
 
             private bool ShouldImplementProperty(string name, Type type)
             {
-                foreach (var interfaceType in _typeBuilder.GetInterfaces())
+                if (_interfaceProperties == null)
                 {
-                    if (interfaceType.GetProperty(name, type) != null)
+                    _interfaceProperties = new HashSet<Tuple<string, Type>>();
+                    foreach (var interfaceType in _typeBuilder.GetInterfaces())
                     {
-                        return true;
+                        // TypeBuilder.GetInterfaces() returns only the interfaces that was explicitly passed to its constructor.
+                        // During compilation the interface hierarchy is flattened, so we only need to resolve one level of ancestral interfaces.
+                        foreach (var parentInterface in interfaceType.GetInterfaces())
+                        {
+                            foreach (var property in parentInterface.GetProperties())
+                            {
+                                _interfaceProperties.Add(new Tuple<string, Type>(property.Name, property.PropertyType));
+                            }
+                        }
+
+                        foreach (var property in interfaceType.GetProperties())
+                        {
+                            _interfaceProperties.Add(new Tuple<string, Type>(property.Name, property.PropertyType));
```

Comment:
What about name conflicts with interface names.?

Reply:
Nice catch - the parser is not supposed to throw an error. Note that this code is currently not running during parsing, but it was initially and could in the future if we can detect all of the errors that are detected by the CLR in constructing a type.

Reply:
It seems there may be two questions.
1. conflict of names of interface properties. I am still not understand what is the interface whose property we implement if signatures is same.
2. conflict with class property names.

Reply:
1) We have neither syntax nor a mechanism in the type definer for _explicitly_ implementing interface members, meaning that multiple interfaces with colliding property signatures will always resolve to the same class property anyways.

2) See 1) 😄